Make Essentials lifecycle registration happen by default

Redth opened the pull request:
Currently we have a few samples that are adding their own startup code to register the right lifecycle events and wire them back up to Essentials'.

Ideally this is done for users when using the MauiAppBuilder, and so this PR brings the good ideas from the MAUI Repo sample into Core itself and automatically wires up the necessary essentials lifecycle events so that the consumer doesn't need to do this.

It also brings the concept of ConfigureEssentials where you can add app actions, enable version tracking, set the maps key, etc.

Redth committed:
Make ConfigureEssentials host builder a thing

Redth committed:
Clean up windows window messages + lifecycle

Redth committed:
Add windows lifecycle for window subclassed

Eilon commented on src/Core/src/Hosting/MauiAppBuilder.cs:
```diff
@@ -56,6 +57,8 @@ internal MauiAppBuilder(bool useDefaults = true)
 				this.ConfigureDispatching();
 			}
+
+			this.UseEssentials();
```
Should this go inside `if (useDefaults)`? Just an open question, I have no horses and this isn't a horse race...

Redth replied:
I had this thought too... Maybe not? but I believe we depend on some things happening here like on android for permissions that some API's will not work if they aren't wired up to the right lifecycle events...

I'm could go either way on this one...

Eilon replied:
Yeah I don't have background on this unfortunately. Another way to look at it, is it "more" or "less" needed than what's already in the `if()`? Or about the same?

Redth replied:
Yea I think I'll move this inside, given that things like Handlers and Dispatching are in there already, this feels about the same level of 'importantness'

Redth committed:
Add OnCreate to Android's lifecycle events which are invoked

Also removes the overrides from the templates since these are now wired up automatically for Essentials

Redth committed:
Fix ordering of base call to OnRequestPermissionsResult

This should come first so that apps have an opportunity to handle this before the base does.

Redth committed:
Remove redundant OnLaunched call in Windows' templates

Redth committed:
Move UseEssentials into the "useDefaults" scope

Redth commented on src/Core/src/Platform/Android/MauiAppCompatActivity.cs:
```diff
@@ -28,6 +29,8 @@ protected override void OnCreate(Bundle? savedInstanceState)
 
 			base.OnCreate(savedInstanceState);
 
+			MauiApplication.Current.Services?.InvokeLifecycleEvents<AndroidLifecycle.OnCreate>(del => del(this, savedInstanceState));
```
Ahh i missed this... but in reality i supect we almost want 2 methods here especially for OnCreate... one at the start, and one at the end... though at the point you need that much control over the order, there's always the subclass...

mattleibow replied:
Actually, reading the docs:

> Called when the Activity calls `super.onCreate()`.

So basically, this line is not needed at all and the event should be firing in the exact same place.

Redth replied:
Yep, removed my accidental duplicate.

Redth committed:
Merge branch 'main' into essentials-lifecycle-registration

Redth committed:
Revert duplicate oncreate lifecycle event invocation

Redth committed:
Merge branch 'main' into essentials-lifecycle-registration

rmarinho commented:
Build is failing 

```
  D:\agent\1\s\src\Core\src\Platform\Windows\MauiWinUIWindow.cs(68,125): error CS1503: Argument 2: cannot convert from 'System.IntPtr' to 'Microsoft.Maui.Platform.WindowsNativeWindowSubclassedEventArgs' [D:\agent\1\s\src\Core\src\Core-net6.csproj]
```

Redth committed:
Merge branch 'main' into essentials-lifecycle-registration

Redth committed:
Fix parameter type
